File: stacks/day2/queue_implementation.py
class Node:
    def __init__(self, val) -> None:
        self.val = val
        self.next = None

class DoublyEndedList:
    def __init__(self) -> None:
        self.front = None
        self.rear = None

    def pushAtRear(self, val):
        node = Node(val)
        if self.isEmpty():
            self.front = node
            self.rear = node
        self.rear.next = node
        self.rear = self.rear.next

    def popAtFront(self):
        if self.isEmpty():
            return
        self.front = self.front.next
        if self.front is None:
            self.rear = None
        
    def isEmpty(self):
        return self.front is None
    
class Queue:
    def __init__(self) -> None:
        self.q = DoublyEndedList()

    def enqueue(self, val):
        self.q.pushAtRear(val)

    def dequeue(self):
        self.q.popAtFront()
    
    def front(self):
        if self.isEmpty(): return None
        return self.q.front.val
    
    def isEmpty(self):
        return self.q.isEmpty()
    

q = Queue()
q.enqueue(10)
q.enqueue(20)
q.enqueue(30)
print(q.front())
q.dequeue()
print(q.front())

File: queues/queues implementation/queue_implementation.cpp
#include <iostream>

class DoublyLinkedList{
    class Node{
        public:
            int val;
            Node * next;
            Node(int val){
                this->val = val;
                this->next = nullptr;
            }
    };
    
    private:
        Node* front;
        Node* rear;
    
    public:
        DoublyLinkedList(){
            this->front = nullptr;
            this->rear = nullptr;
        }

        void pushAtRear(int val){
            Node* newNode = new Node(val);
            if(this->front == nullptr){
                this->front = newNode;
                this->rear = newNode;
            }
            else{
                this->rear->next = newNode;
                this->rear = newNode;
            }
        }

        int frontVal(){
            if(this->front == nullptr){
                return -1;
            }
            else{
                return this->front->val;
            }
        }

        void popAtFront(){
            if(this->front != nullptr){
                Node* temp = this->front;
                this->front = this->front->next;
                delete temp;
                if(this->front == nullptr){
                    this->rear = nullptr;
                }
            }
        }

        bool isEmpty(){
            return this->front == nullptr;
        }
};

class Queue{
    private: 
        DoublyLinkedList* q;
    
    public:
        Queue(){
            this->q = new DoublyLinkedList();
        }

        ~Queue(){
            delete this->q;
        }

        void enqueue(int val){
            this->q->pushAtRear(val);
        }

        int front(){
            if(this->q->isEmpty()) return -1;
            else return this->q->frontVal();
        }

        void dequeue(){
            this->q->popAtFront();
        }

        bool isEmpty(){
            return this->q->isEmpty();
        }
};

int main(){
    Queue q;
    q.enqueue(10);
    q.enqueue(20);
    q.enqueue(30);

    std::cout << q.front() << std::endl;
    q.dequeue();
    std::cout << q.front() << std::endl;

    return 0;
}
